subsys/logging/log_backend_swo.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2018 Piotr Mienkowski
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ * @brief Serial Wire Output (SWO) backend implementation.
+ *
+ * SWO/SWV has been developed by ARM. The following code works only on ARM
+ * architecture.
+ *
+ * An SWO viewer program will typically set-up the SWO port including its
+ * frequency when connected to the debug probe. Such configuration can persist
+ * only until the MCU reset. The SWO backend initialization function will
+ * re-configure the SWO port upon boot and set the frequency as specified by
+ * the LOG_BACKEND_SWO_FREQ_HZ Kconfig option. To ensure flawless operation
+ * this frequency should much the one set by the SWO viewer program.
+ *
+ * The initialization code assumes that SWO core frequency is equal to HCLK
+ * as defined by SYS_CLOCK_HW_CYCLES_PER_SEC Kconfig option. This may require
+ * additional, vendor specific configuration.
+ */
+
+#include <logging/log_backend.h>
+#include <logging/log_core.h>
+#include <logging/log_msg.h>
+#include <logging/log_output.h>
+#include <soc.h>
+
+/** The stimulus port from which SWO data is received and displayed */
+#define ITM_PORT_LOGGER       0
+
+/* Set TPIU prescaler for the current debug trace clock frequency. */
+#if CONFIG_LOG_BACKEND_SWO_FREQ_HZ == 0
+#define SWO_FREQ_DIV  1
+#else
+#define SWO_FREQ (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC \
+		  + (CONFIG_LOG_BACKEND_SWO_FREQ_HZ / 2))
+#define SWO_FREQ_DIV  (SWO_FREQ / CONFIG_LOG_BACKEND_SWO_FREQ_HZ)
+#if SWO_FREQ_DIV > 0xFFFF
```

Comment:
it could be changed to `BUILD_ASSERT_MSG` or `#error`. I haven't yet seen such pragma in zephyr.

Reply:
`#pragma message` is used as a missing `#warning` directive. In this case `#error` is indeed more appropriate. I've made the change.